---
title: "Power calculation"
output: html_document
---

```{r}
library(data.table)
library(MASS)
library(ggplot2)
```

```{r}
reads <- fread("../data/reads.293FTDR_S1_R1_001.fastq_bq10_star_corrected.umi.dge.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE)
```

```{r}
total_reads_per_sample <- apply(as.matrix(reads[,-1]),2, sum)
hist(total_reads_per_sample,main = "Distribution of reads number across samples")
total_reads_per_gene_across_samples <- apply(as.matrix(reads[,-1]),1, mean)
hist(log(total_reads_per_gene_across_samples + 0.01,2),main = "Distribution of mean reads per gene across samples")
quantile(total_reads_per_gene_across_samples)
reads_without_genename <- reads[,-1]
reads_gt20000 <- reads_without_genename[, .SD, .SDcols = as.vector(total_reads_per_sample >= 20000)]
# There are 81 samples with more than 20K reads for each cell. Will use these data to do downsampling and estimate dispersion.
genes <- reads[[1]]
```


Need to come up a function to random sample 20000 reads for each cell with more than 20000 reads. 
```{r}
# Notice, the theta estimation fails for some genes, and this function will record the estimate as NA in the output
random_sample_for_one_sample <- function(vector, n, genes){
  # [vector] is a vector with read counts per gene
  # [n] is the total number of sampled read counts
  # [genes] a vector of gene names, the order should be consistent with the read counts in [vector]
  temp <- data.frame(genes = genes, count = vector)
  temp <- temp[temp$count != 0,]
  # as.vector(temp[,1]) is much much faster than using temp[,1] directly
  all_reads_per_sample <- as.vector(unlist(mapply(function(x,y) rep(x,y), as.vector(temp[,1]), as.vector(temp[,2]), SIMPLIFY = TRUE)))
  sampled_reads <- sample(all_reads_per_sample, n)
  sampled_reads <- as.data.frame(table(sampled_reads))
  colnames(sampled_reads) <- c("genename", "count_sampled")
  all_genes_reads_holder <- data.frame(genename = genes)
  all_genes_reads_holder$id = seq(1,length(all_genes_reads_holder[,1]))
  all_genes_reads_holder <- merge(all_genes_reads_holder, sampled_reads, by = "genename", all.x = TRUE)
  all_genes_reads_holder[is.na(all_genes_reads_holder$count_sampled),]$count_sampled <- 0
  all_genes_reads_holder <- all_genes_reads_holder[order(all_genes_reads_holder$id, decreasing = FALSE),]
  return(all_genes_reads_holder$count_sampled)
}


dispersion_calculation <- function(target_gene, count_matrix, genes){
  #[count_matrix] is a matrix that has samples with down-sampled read counts
  #[genes] is a vector of gene names, the order should be consistent with the read counts in [count_matrix]
  #[target_gene] is the gene that we are going to use to estimate the dispersion parameter.
  data <- data.frame(count = count_matrix[genes == target_gene,])
  tryCatch(eval(m1 <- glm.nb(count ~ 1, data = data, control=glm.control(maxit=100000))), error = function(e) message("Oops!  ", as.character(e)))
  if(exists("m1")){
    return(summary(m1)$theta)
  }
  else{
    return(NA)
  }
}
```

```{r, cache = TRUE, dependson=random_sample_for_one_sample, dependson=dispersion_calculation, dependson=reads_gt20000}

#sample 20000 reads from each sample
set.seed(345)
t <- apply(reads_gt20000, 2, random_sample_for_one_sample, n = 20000, genes = genes)
quantile(apply(t,1,mean), seq(0,1,0.01))
rownames(t) <- genes
#t[apply(t,1,mean)> 1.91358024 & apply(t,1,mean) < 1.91358026,]
# MRPL35 top 90%, mean reads about 2
#t[apply(t,1,mean)> 5.1 & apply(t,1,mean) < 5.2,]
# SEC62 top 97%, mean reads number about 12 about 5
#t[apply(t,1,mean)> 12.1 & apply(t,1,mean) < 13,]
# MYL6 top 99% mean reads number about 12
dispersion_calculation("MRPL35", t, genes)
dispersion_calculation("NUB1", t, genes)
dispersion_calculation("SEC62", t, genes)
dispersion_calculation("MYL6", t, genes)

genes_below1 <- rownames(t[apply(t,1,mean)<1,])
genes_1_to_5 <- rownames(t[apply(t,1,mean) >=1 & apply(t,1,mean) < 5,])
genes_5_to_10 <- rownames(t[apply(t,1,mean) >=5 & apply(t,1,mean) < 10,])
genes_above10 <- rownames(t[apply(t,1,mean) >=10,])

temp <- c()
hist(mapply(dispersion_calculation, genes_above10, MoreArgs = list(count_matrix = t, genes = genes)), main = "Genes with mean reads greater than 10")

summary(mapply(dispersion_calculation, genes_above10, MoreArgs = list(count_matrix = t, genes = genes)))

hist(mapply(dispersion_calculation, genes_5_to_10, MoreArgs = list(count_matrix = t, genes = genes)), main = "Genes with mean reads between 5 to 10")

summary(mapply(dispersion_calculation, genes_5_to_10, MoreArgs = list(count_matrix = t, genes = genes))) 

hist(mapply(dispersion_calculation, genes_1_to_5, MoreArgs = list(count_matrix = t, genes = genes)), main = "Genes with mean reads between 1 to 5")

summary(mapply(dispersion_calculation, genes_1_to_5, MoreArgs = list(count_matrix = t, genes = genes))) 

 
hist(mapply(dispersion_calculation, genes_below1, MoreArgs = list(count_matrix = t, genes = genes)), main = "Genes with mean reads smaller than 1")

summary(mapply(dispersion_calculation, genes_below1, MoreArgs = list(count_matrix = t, genes = genes))) 


```






Simulation will be performed assume thare are 50K reads for each cell. 

```{r}

test <- rnbinom(1000,mu = 30, size = 1.2)
test <- data.frame(outcome = test)
m1 <- glm.nb(outcome ~ 1, data = test)
summary(m1)

```


```{r}
generate_single_cell_data <- function(sample_size = 200, effect_size = 2, reads = 10, theta = 10, simulation_rounds = 100, BF.correction = 40){
  #[sample_size] is the number of cells in treatment group or control group
  # [effect size] is the effect of treatment on RNA expression in fold of increase
  # [reads] is the mean reads count in the control group
  # [theta] is the shape parameter for the NB distribution
  dummy <- c()
  for(i in 1:simulation_rounds){
    treatment <- rnbinom(sample_size,mu = reads * effect_size, size = theta)
    control <- rnbinom(sample_size,mu = reads, size = theta)
    data <- data.frame(count = c(treatment, control), class = c(rep("treatment", sample_size), rep("control", sample_size)))
    tryCatch(eval(m1 <- glm.nb(count ~ class, data = data, control=glm.control(maxit=1000000))), error = function(e) message("Oops!  ", as.character(e)))
    if(exists("m1")){
      if(summary(m1)$coef[2,4]< (0.05/BF.correction)){
        dummy[i] <- 1
      }else{
        dummy[i] <- 0
      }
    }
    else{
      dummy[i] <- 0
    }
  }
  return(mean(dummy))
}

```


# 10 reads on average in control

```{r}
sample_size_gradient <- c(10, 50, 150, 200)
effect_size_gradient <- c(1.25, 1.5, 1.75, 2)
report_for_10_reads <- data.frame(sample_size = rep(sample_size_gradient, 4), effect_size = rep(effect_size_gradient, each = 4), power = 0)
for(i in 1:length(report_for_10_reads[,1])){
  report_for_10_reads[i,]$power <- generate_single_cell_data(sample_size = report_for_10_reads[i,]$sample_size, effect_size = report_for_10_reads[i,]$effect_size, reads = 10, theta = 10, simulation_rounds = 100, BF.correction = 40)
}

report_for_10_reads$effect_size <- as.factor(report_for_10_reads$effect_size)
levels(report_for_10_reads$effect_size) <- c("1.25", "1.5", "1.75", "2")

```


```{r}
p <- ggplot(data = report_for_10_reads, aes(x = sample_size, y = power, group = effect_size))
p <- p + geom_line(aes(linetype = effect_size, color = effect_size))+ geom_point(aes(color = effect_size)) + theme(legend.position="top")
p
```

# 5 reads on average in control


```{r}
report_for_10_reads <- data.frame(sample_size = rep(sample_size_gradient, 4), effect_size = rep(effect_size_gradient, each = 4), power = 0)
for(i in 1:length(report_for_10_reads[,1])){
  report_for_10_reads[i,]$power <- generate_single_cell_data(sample_size = report_for_10_reads[i,]$sample_size, effect_size = report_for_10_reads[i,]$effect_size, reads = 5, theta = 10, simulation_rounds = 100, BF.correction = 40)
}

report_for_10_reads$effect_size <- as.factor(report_for_10_reads$effect_size)
levels(report_for_10_reads$effect_size) <- c("1.25", "1.5", "1.75", "2")

```


```{r}
p <- ggplot(data = report_for_10_reads, aes(x = sample_size, y = power, group = effect_size))
p <- p + geom_line(aes(linetype = effect_size, color = effect_size))+ geom_point(aes(color = effect_size)) + theme(legend.position="top")
p
```


# 1 reads on average in control


```{r, eval=FALSE}
report_for_10_reads <- data.frame(sample_size = rep(sample_size_gradient, 4), effect_size = rep(effect_size_gradient, each = 4), power = 0)
for(i in 1:length(report_for_10_reads[,1])){
  report_for_10_reads[i,]$power <- generate_single_cell_data(sample_size = report_for_10_reads[i,]$sample_size, effect_size = report_for_10_reads[i,]$effect_size, reads = 1, theta = 10, simulation_rounds = 100, BF.correction = 40)
}

report_for_10_reads$effect_size <- as.factor(report_for_10_reads$effect_size)
levels(report_for_10_reads$effect_size) <- c("1.25", "1.5", "1.75", "2")

```


```{r, eval=FALSE}
p <- ggplot(data = report_for_10_reads, aes(x = sample_size, y = power, group = effect_size))
p <- p + geom_line(aes(linetype = effect_size, color = effect_size))+ geom_point(aes(color = effect_size)) + theme(legend.position="top")
p
```